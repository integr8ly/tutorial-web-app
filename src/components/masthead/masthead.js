import React from 'react';
import PropTypes from 'prop-types';
import {
  Brand,
  Button,
  Dropdown,
  DropdownToggle,
  DropdownItem,
  DropdownSeparator,
  PageHeader,
  Toolbar,
  ToolbarGroup,
  ToolbarItem
} from '@patternfly/react-core';
import { CogIcon, HelpIcon } from '@patternfly/react-icons';
import accessibleStyles from '@patternfly/patternfly/utilities/Accessibility/accessibility.css';
import { css } from '@patternfly/react-styles';
import { withRouter } from 'react-router-dom';
import { noop } from '../../common/helpers';
import { connect, reduxActions } from '../../redux';
import { AboutModal } from '../aboutModal/aboutModal';
import { logout } from '../../services/openshiftServices';
import solutionExplorerImg from '../../img/Logo-Solution-Explorer-Reverse-RGB.svg';
import managedIntegrationSolutionExplorerImg from '../../img/Logo-Red-Hat-Managed-Integration-Solution-Explorer-Reverse-RGB.svg';

class Masthead extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isHelpDropdownOpen: false,
      isUserDropdownOpen: false,
      showAboutModal: false
    };

    this.onTitleClick = this.onTitleClick.bind(this);
    this.onLogoutUser = this.onLogoutUser.bind(this);

    this.onUserDropdownToggle = this.onUserDropdownToggle.bind(this);
    this.onUserDropdownSelect = this.onUserDropdownSelect.bind(this);

    this.onHelpDropdownToggle = this.onHelpDropdownToggle.bind(this);
    this.onHelpDropdownSelect = this.onHelpDropdownSelect.bind(this);

    this.onAboutModal = this.onAboutModal.bind(this);
    this.closeAboutModal = this.closeAboutModal.bind(this);
  }

  onLogoutUser = () => {
    if (window.OPENSHIFT_CONFIG.mockData) {
      window.localStorage.clear();
      window.location.href = '/';
      return;
    }
    logout().then(() => {
      window.location.href = window.OPENSHIFT_CONFIG.ssoLogoutUri;
    });
  };

  onAboutModal(e) {
    e.preventDefault();
    this.setState({ showAboutModal: true });
  }

  closeAboutModal() {
    this.setState({ showAboutModal: false });
  }

  onTitleClick = () => {
    const { history } = this.props;
    history.push('/');
  };

  onUserDropdownToggle(isUserDropdownOpen) {
    this.setState({
      isUserDropdownOpen
    });
  }

  onUserDropdownSelect = () => {
    this.setState({
      isUserDropdownOpen: !this.state.isUserDropdownOpen
    });
  };

  onSettingsClick = () => {
    const { history } = this.props;
    history.push(`/settings`);
  };

  onDevResourcesClick = () => {
    const { history } = this.props;
    history.push(`/dev-resources`);
  };

  onHelpDropdownToggle(isHelpDropdownOpen) {
    this.setState({
      isHelpDropdownOpen
    });
  }

  getLogo = () => {
    let clusterType = '';
    let logoName = '';
    if (window.OPENSHIFT_CONFIG) {
      clusterType = window.OPENSHIFT_CONFIG.mockData ? 'localhost' : window.OPENSHIFT_CONFIG.clusterType;
      if (clusterType === 'poc') {
        logoName = managedIntegrationSolutionExplorerImg;
      } else if (clusterType === 'osd') {
        logoName = managedIntegrationSolutionExplorerImg;
      } else {
        logoName = solutionExplorerImg;
      }
    }
    return logoName;
  };

  onHelpDropdownSelect = () => {
    this.setState({
      isHelpDropdownOpen: !this.state.isHelpDropdownOpen
    });
  };

  render() {
    const { isUserDropdownOpen, isHelpDropdownOpen, showAboutModal } = this.state;

    const logoProps = {
      onClick: () => this.onTitleClick()
    };

    const gsUrl =
      'https://access.redhat.com/documentation/en-us/red_hat_managed_integration/1/html-single/getting_started/';
    const riUrl =
      'https://access.redhat.com/documentation/en-us/red_hat_managed_integration/1/html-single/release_notes/';
    const csUrl = 'https://access.redhat.com/support/';

    const MastheadToolbar = (
      <React.Fragment>
        <Toolbar>
          <ToolbarGroup className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnLg)}>
            <ToolbarItem className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnMd)}>
              <Button
                className="pf-c-button pf-m-plain"
                aria-label="Settings"
                variant="plain"
                onClick={this.onSettingsClick}
              >
                <CogIcon />
              </Button>
            </ToolbarItem>
            <ToolbarItem className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnMd)}>
              <Dropdown
                isPlain
                position="right"
                onSelect={this.onHelpDropdownSelect}
                isOpen={isHelpDropdownOpen}
                toggle={
                  <DropdownToggle
                    iconComponent={null}
                    onToggle={this.onHelpDropdownToggle}
                    aria-label="Link to Help page"
                  >
                    <HelpIcon />
                  </DropdownToggle>
                }
                autoFocus={false}
                dropdownItems={[
                  <DropdownItem
                    key="help-getting-started"
                    href={gsUrl}
                    target="_blank"
                    aria-label="Link to getting started page"
                  >
                    Getting started
                  </DropdownItem>,
                  <DropdownItem
                    key="help-release-info"
                    href={riUrl}
                    target="_blank"
                    aria-label="Link to release information page"
                  >
                    Release information
                  </DropdownItem>,
                  <DropdownItem
                    key="help-customer-support"
                    href={csUrl}
                    target="_blank"
                    aria-label="Link to customer support page"
                  >
                    Customer support
                  </DropdownItem>,
                  <DropdownSeparator key="help-separator-1" />,
                  <DropdownItem key="help-dev-resources" onClick={this.onDevResourcesClick}>
                    Developer resources
                  </DropdownItem>,
                  <DropdownSeparator key="help-separator-2" />,
                  <DropdownItem
                    key="about"
                    component="button"
                    href="#about"
                    onClick={this.onAboutModal}
                    aria-label="About"
                  >
                    About
                  </DropdownItem>
                ]}
              />
            </ToolbarItem>
          </ToolbarGroup>
          <ToolbarGroup className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnSm)}>
            <ToolbarItem className={css(accessibleStyles.screenReader, accessibleStyles.visibleOnSm)}>
              <Dropdown
                isPlain
                position="right"
                onSelect={this.onUserDropdownSelect}
                isOpen={isUserDropdownOpen}
                toggle={
                  <DropdownToggle onToggle={this.onUserDropdownToggle}>
                    {window.localStorage.getItem('currentUserName')}
                  </DropdownToggle>
                }
                autoFocus={false}
                dropdownItems={[
                  <DropdownItem
                    key="logout"
                    component="button"
                    href="#logout"
                    onClick={this.onLogoutUser}
                    aria-label="Log out of the system"
                  >
                    Log out
                  </DropdownItem>
                ]}
              />
            </ToolbarItem>
          </ToolbarGroup>
        </Toolbar>
        {showAboutModal && <AboutModal isOpen={showAboutModal} closeAboutModal={this.closeAboutModal} />}
      </React.Fragment>
    );

    return (
      <PageHeader
        logo={<Brand src={this.getLogo()} alt="Red Hat Solution Explorer" />}
        logoProps={logoProps}
        toolbar={MastheadToolbar}
      />
    );
  }
}

Masthead.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  })
};

Masthead.defaultProps = {
  history: {
    push: noop
  }
};

const mapDispatchToProps = dispatch => ({
  logoutUser: () => dispatch(reduxActions.userActions.logoutUser())
});

const ConnectedMasthead = connect(
  undefined,
  mapDispatchToProps
)(Masthead);

const RoutedConnectedMasthead = withRouter(ConnectedMasthead);

export { RoutedConnectedMasthead as default, ConnectedMasthead, RoutedConnectedMasthead, Masthead };
