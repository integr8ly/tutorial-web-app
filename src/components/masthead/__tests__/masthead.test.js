import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow, mount } from 'enzyme';
import { ConnectedMasthead, Masthead } from '../masthead';

describe('Masthead Component', () => {
  const generateEmptyStore = (obj = {}) => configureMockStore()(obj);

  it('should render a basic component', () => {
    const store = generateEmptyStore({ user: { session: { username: 'Admin' } } });
    const component = shallow(<ConnectedMasthead />, { context: { store } });

    expect(component).toMatchSnapshot('basic');
    expect(component.dive().find('#app-help-dropdown')).toMatchSnapshot('help dropdown');
    expect(component.dive().find('#app-user-dropdown')).toMatchSnapshot('user dropdown');
  });

  it('should render mobile navigation', () => {
    const props = {};
    const component = mount(<Masthead {...props} />);
    const componentInstance = component.instance();

    component.setState({ mobileToggle: false });
    expect(component.state().mobileToggle).toEqual(false);
    expect(component.find('div[role="menu"]')).toMatchSnapshot('mobile-nav');
    componentInstance.navToggle();
    expect(component.state().mobileToggle).toEqual(true);
  });

  it('should have specific events defined', () => {
    const component = mount(<Masthead />);
    const componentInstance = component.instance();

    expect(componentInstance.onAbout).toBeDefined();
    expect(componentInstance.onHelp).toBeDefined();
    expect(componentInstance.onLogoutUser).toBeDefined();
  });

  it('should handle basic events', () => {
    const logoutUser = jest.fn();
    const props = {
      logoutUser
    };
    const component = mount(<Masthead {...props} />);
    const componentInstance = component.instance();
    const mockEvent = { preventDefault: () => {} };

    componentInstance.onLogoutUser(mockEvent);
    expect(logoutUser).toHaveBeenCalled();
  });
});
