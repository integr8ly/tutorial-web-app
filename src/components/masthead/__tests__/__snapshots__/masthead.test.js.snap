// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Masthead Component should render a basic component: basic 1`] = `
<Masthead
  history={Object {}}
  logoutUser={[Function]}
  user={
    Object {
      "username": "Admin",
    }
  }
/>
`;

exports[`Masthead Component should render a basic component: help dropdown 1`] = `
<MastheadDropdown
  className=""
  id="app-help-dropdown"
  noCaret={true}
  title={
    <span
      aria-hidden={true}
      className="pficon pficon-help"
    />
  }
>
  <MenuItem
    bsClass="dropdown"
    disabled={false}
    divider={false}
    eventKey="1"
    header={false}
    onClick={[Function]}
  >
    Help
  </MenuItem>
  <MenuItem
    bsClass="dropdown"
    disabled={false}
    divider={false}
    eventKey="2"
    header={false}
    onClick={[Function]}
  >
    About
  </MenuItem>
</MastheadDropdown>
`;

exports[`Masthead Component should render a basic component: user dropdown 1`] = `
<MastheadDropdown
  className=""
  id="app-user-dropdown"
  noCaret={false}
  title={
    <React.Fragment>
      <Icon
        name="user"
        type="pf"
      />
       
      <span
        className="dropdown-title"
      >
        Admin
         
         
      </span>
    </React.Fragment>
  }
>
  <MenuItem
    bsClass="dropdown"
    disabled={false}
    divider={false}
    header={false}
    onClick={[Function]}
  >
    Logout
  </MenuItem>
</MastheadDropdown>
`;

exports[`Masthead Component should render mobile navigation: mobile-nav 1`] = `
<div
  aria-live="polite"
  className="nav-pf-vertical nav-pf-vertical-with-sub-menus nav-pf-vertical-with-badges hidden show-mobile-nav"
  role="menu"
>
  <ul
    className="list-group"
  >
    <li
      className="list-group-item"
    >
      <a
        href="#about"
        onClick={[Function]}
        role="menuitem"
      >
        <span
          className="list-group-item-value"
        >
          About
        </span>
      </a>
    </li>
    <li
      className="list-group-item"
    >
      <a
        href="#help"
        onClick={[Function]}
        role="menuitem"
      >
        <span
          className="list-group-item-value"
        >
          Help
        </span>
      </a>
    </li>
    <li
      className="list-group-item"
    >
      <a
        href="#logout"
        onClick={[Function]}
        role="menuitem"
      >
        <span
          className="list-group-item-value"
        >
          Logout
        </span>
      </a>
    </li>
  </ul>
</div>
`;
